src/collector/sort_key/sort_key_computer.rs

```diff
@@ -0,0 +1,361 @@
+use std::cmp::Ordering;
+
+use crate::collector::sort_key::ReverseOrder;
+use crate::{DocId, Order, Result, Score, SegmentReader};
+
+/// A `SegmentSortKeyComputer` makes it possible to modify the default score
+/// for a given document belonging to a specific segment.
+///
+/// It is the segment local version of the [`SortKeyComputer`].
+pub trait SegmentSortKeyComputer: 'static {
+    /// The final score being emitted.
+    type SortKey: 'static + PartialOrd + Send + Sync + Clone;
+
+    /// Sort key used by at the segment level by the `SegmentSortKeyComputer`.
+    ///
+    /// It is typically small like a `u64`, and is meant to be converted
+    /// to the final score at the end of the collection of the segment.
+    type SegmentSortKey: 'static + PartialOrd + Clone + Send + Sync + Clone;
+
+    /// Computes the sort key for the given document and score.
+    fn sort_key(&mut self, doc: DocId, score: Score) -> Self::SegmentSortKey;
+
+    /// Returns true if the `SegmentSortKeyComputer` is a good candidate for the lazy evaluation
+    /// optimization. See [`SegmentSortKeyComputer::accept_score_lazy`].
+    fn is_lazy() -> bool {
+        false
+    }
+
+    /// Implementing this method makes it possible to avoid computing
+    /// a sort_key entirely if we can assess that it won't pass a threshold
+    /// with a partial computation.
+    ///
+    /// This is currently used for lexicographic sorting.
+    ///
+    /// If REVERSE_ORDER is false (resp. true),
+    /// - we return None if the score is below the threshold (resp. above to the threshold)
+    /// - we return Some(ordering, score) if the score is above or equal to the threshold (resp.
+    ///   below or equal to)
+    fn accept_sort_key_lazy<const REVERSE_ORDER: bool>(
+        &mut self,
+        doc_id: DocId,
+        score: Score,
+        threshold: &Self::SegmentSortKey,
+    ) -> Option<(std::cmp::Ordering, Self::SegmentSortKey)> {
+        let excluded_ordering = if REVERSE_ORDER {
+            Ordering::Greater
+        } else {
+            Ordering::Less
+        };
```

Comment:
Based on some testing (hardcoding `is_lazy = true` and then increasing the testcase size of `test_top_field_collect_string_prop`), these cases are reversed.

Lifting over some of the tests from #2681 might be a good way to get some coverage of the multi-sort key cases quickly.